x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/QueryBuilderResolver.java

```diff
@@ -0,0 +1,139 @@
+            return;
+        }
+        ConcurrentMap<FullTextFunction, QueryBuilder> results = new ConcurrentHashMap<>();
+
+        // We are using a GroupedActionListener here so that we update the plan only after al the QueryBuilders have been resolved.
+        // This way we avoid making concurrent updates to the plan.
+        GroupedActionListener<QueryBuilder> groupedActionListener = new GroupedActionListener<>(
+            unresolved.size(), // how many QueryBuilders need to be collected before onResponse is called
+            new ActionListener<Collection<QueryBuilder>>() {
+                @Override
+                public void onResponse(Collection<QueryBuilder> ignored) {
+                    // This is called only after the groupedActionListener has received all the QueryBuilders
+                    try {
+                        LogicalPlan newPlan = planWithResolvedQueryBuilders(plan, results);
+                        callback.accept(newPlan, listener);
+                    } catch (Exception e) {
+                        onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            }
+        );
+
+        QueryRewriteContext ctx = queryRewriteContext(indexNames);
+
+        for (FullTextFunction fullTextFunction : unresolved) {
+            // First we get the initial query builder
+            QueryBuilder queryBuilder = fullTextFunction.asQuery(PlannerUtils.TRANSLATOR_HANDLER).asBuilder();
+            // Then the rewrite will produce a new query builder
+            Rewriteable.rewriteAndFetch(queryBuilder, ctx, groupedActionListener.delegateFailureAndWrap((next, resolvedQueryBuilder) -> {
+                results.put(fullTextFunction, resolvedQueryBuilder);
+                next.onResponse(resolvedQueryBuilder);
+            }));
+        }
+    }
+
+    private Set<FullTextFunction> fullTextFunctions(LogicalPlan plan) {
+        Set<FullTextFunction> functions = new HashSet<>();
+        plan.forEachExpressionDown(FullTextFunction.class, func -> functions.add(func));
+        return functions;
+    }
+
+    public Set<String> indexNames(LogicalPlan plan) {
+        Holder<Set<String>> indexNames = new Holder<>();
+
+        plan.forEachDown(EsRelation.class, esRelation -> { indexNames.set(esRelation.index().concreteIndices()); });
+
+        return indexNames.get();
+    }
+
+    public LogicalPlan planWithResolvedQueryBuilders(LogicalPlan plan, ConcurrentMap<FullTextFunction, QueryBuilder> newQueryBuilders) {
+        LogicalPlan newPlan = plan.transformExpressionsDown(FullTextFunction.class, m -> {
+            if (newQueryBuilders.keySet().contains(m)) {
+                return m.replaceQueryBuilder(newQueryBuilders.get(m));
+            }
+            return m;
+        });
+        newPlan.setAnalyzed();
+        newPlan.setOptimized();
```

Comment:
oh, this is a bit esoteric ;-) Maybe add a comment/assert that it preserves the original plan state here.

Reply:
Yes - we need the plan to be optimized, I will make it explicit.
I am getting some errors otherwise, for example for:

```
    FROM testidx
    | RENAME semantic_text AS a 
    | WHERE match(a, "live")
```
It complains with `Match must have a field attribute as the first argument` - this works when the plan is optimized. This is because we see the field as a `ReferenceAttribute` rather than a `FieldAttribute`.
I could potentially try and fix this in the `MatchFunctionTranslator`, but that means we always have to be careful when we add a new function, that the translator also checks for `ReferenceAttribute`.

There is also the advantage that during optimization stage some of the filter conditions might be dropped.
If we don't do the optimization stage before, we might generate `QueryBuilder`s for `FullTextFunction` objects that later on might be dropped.